sunscreen/benches/fractional_range_proof.rs

```diff
@@ -1,9 +1,141 @@
+use std::time::Instant;
+
 use criterion::{criterion_group, criterion_main, Criterion};
-use sunscreen::{types::zkp::NativeField, *};
+use sunscreen::{
+    types::zkp::{ConstrainCmp, IntoProgramNode, NativeField, ProgramNode},
+    *,
+};
+use sunscreen_zkp_backend::{bulletproofs::BulletproofsBackend, BigInt};
+
+type BPField = <BulletproofsBackend as ZkpBackend>::Field;
 
 fn fractional_range_proof(_c: &mut Criterion) {
+    env_logger::init();
+
+    fn to_field_element<F: BackendField>(
+        bits: &[ProgramNode<NativeField<F>>],
+        twos_complement: bool,
+    ) -> ProgramNode<NativeField<F>> {
+        let powers = (0..bits.len())
+            .map(|x| {
+                let power = NativeField::<F>::from(BigInt::from(BigInt::ONE.shl_vartime(x)));
+                if twos_complement && x == bits.len() - 1 {
+                    -(power.into_program_node())
+                } else {
+                    power.into_program_node()
+                }
+            })
+            .collect::<Vec<_>>();
+        let mut val = NativeField::from(0u8).into_program_node();
+
+        for (i, bit) in bits.iter().enumerate() {
+            val = val + *bit * powers[i];
+        }
+
+        val
+    }
+
     #[zkp_program(backend = "bulletproofs")]
-    fn in_range<F: BackendField>(a: [NativeField<F>; 8]) {}
+    /**
+     * Proves the 0 < a <= b and a == c
+     */
+    fn in_range<F: BackendField>(
+        a: [[NativeField<F>; 8]; 64],
+        b: [[NativeField<F>; 8]; 64],
+        c: [[NativeField<F>; 8]; 64],
+    ) {
+        let a_coeffs = a
+            .iter()
+            .map(|x| to_field_element(x, true))
+            .collect::<Vec<_>>();
+        let b_coeffs = b
+            .iter()
+            .map(|x| to_field_element(x, true))
+            .collect::<Vec<_>>();
+        let c_coeffs = c
+            .iter()
+            .map(|x| to_field_element(x, true))
+            .collect::<Vec<_>>();
+
+        let a_val = to_field_element(&a_coeffs, false);
+        let b_val = to_field_element(&b_coeffs, false);
+        let c_val = to_field_element(&c_coeffs, false);
+
+        a_val.constrain_gt_bounded(NativeField::<F>::from(0).into_program_node(), 8);
+        a_val.constrain_le_bounded(b_val, 8);
+        a_val.constrain_eq(c_val);
+    }
+
+    let app = Compiler::new()
+        .zkp_backend::<BulletproofsBackend>()
+        .zkp_program(in_range)
+        .compile()
+        .unwrap();
+
+    let prog = app.get_zkp_program(in_range).unwrap();
+
+    fn encode(val: i8) -> [NativeField<BPField>; 8] {
+        let as_u8 = val.to_le_bytes()[0];
+
+        (0..8)
+            .map(|x| NativeField::from((as_u8 >> x) & 0x1))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap()
+    }
+
+    // 3 * 1 + 2 * 2 = 7
```

Comment:
For this comment (and the other two), could expand on what this comment is for. Or remove it.